src/Compilers/Core/Portable/PEWriter/BlobBuilder.Enumerators.cs

```diff
@@ -0,0 +1,104 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Diagnostics;
+using System.Collections;
+using System.Collections.Generic;
+
+namespace Microsoft.Cci
+{
+    partial class BlobBuilder
+    {
+        protected struct Chunks : IEnumerable<BlobBuilder>, IEnumerator<BlobBuilder>, IEnumerator
+        {
+            private readonly BlobBuilder _last;
+            private readonly BlobBuilder _first;
+            private BlobBuilder _currentOpt;
+
+            internal Chunks(BlobBuilder builder)
+            {
+                Debug.Assert(builder.IsHead);
+
+                _last = builder;
+                _first = builder._nextOrPrevious._nextOrPrevious;
+                _currentOpt = null;
+            }
+
+            object IEnumerator.Current => Current;
+            public BlobBuilder Current => _currentOpt;
+
+            public bool MoveNext()
+            {
+                if (_currentOpt == null)
+                {
+                    _currentOpt = _first;
+                    return true;
+                }
+
+                if (_currentOpt == _last)
+                {
+                    return false;
+                }
+
+                if (_currentOpt == _last._nextOrPrevious)
+                {
+                    _currentOpt = _last;
+                    return true;
+                }
+
+                _currentOpt = _currentOpt._nextOrPrevious;
+                return true;
+            }
+
+            public void Reset()
+            {
+                _currentOpt = null;
+            }
+
+            void IDisposable.Dispose() { }
+
+            // IEnumerable:
+            public Chunks GetEnumerator() => this;
+            IEnumerator<BlobBuilder> IEnumerable<BlobBuilder>.GetEnumerator() => GetEnumerator();
+            IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
+        }
+
+        public struct Blobs : IEnumerable<BlobWriter>, IEnumerator<BlobWriter>, IEnumerator
```

Comment:
I'm going to replace BlobWriter here by Blob, which is going to be an immutable struct wrapping a byte[] (aka slice).